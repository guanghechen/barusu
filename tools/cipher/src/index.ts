export * from './util/buffer'
export * from './util/catalog'
export * from './util/cipher'
export * from './util/error'
export * from './util/event-bus'
export * from './util/io'
export * from './util/logger'
export * from './util/master'
export * from './util/secret'
