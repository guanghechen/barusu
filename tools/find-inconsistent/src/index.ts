import fs from 'fs-extra'
import glob from 'glob'
import path from 'path'
import './env/constant'
import { logger } from './env/logger'
import { checkFatalError } from './util'
export * from './env/constant'
export * from './env/logger'
export * from './util'

export type PackageDependencies = { [name: string]: string }
export type DependencyItem = {
  version: string
  dependents: string[]
}

export interface PackageJSON {
  name: string
  version: string
  bundledDependencies?: PackageDependencies
  dependencies?: PackageDependencies
  devDependencies?: PackageDependencies
  optionalDependencies?: PackageDependencies
  peerDependencies?: PackageDependencies
}

export class PackageManager {
  protected readonly packageMap: Map<string, string> = new Map()
  protected readonly dependencyMap: Map<string, DependencyItem> = new Map()

  public async resolve(rootPackageJsonPath: string): Promise<void> {
    if (!fs.existsSync(rootPackageJsonPath)) {
      throw new Error(`${rootPackageJsonPath} is not found.`)
    }

    const self = this
    self.packageMap.clear()
    self.dependencyMap.clear()
    const manifest = fs.readJSONSync(rootPackageJsonPath)
    self.resolvePackageJSON(manifest)

    if (manifest.workspaces == null) {
      throw new Error(
        `workspaces field is not specified in ${rootPackageJsonPath}`,
      )
    }

    const tasks: Promise<void>[] = []
    for (const workspace of manifest.workspaces) {
      const task = new Promise<void>((resolve, reject) => {
        glob(workspace, (err: any, files: string[]) => {
          if (err) {
            reject(err)
            return
          }

          for (const filePath of files) {
            const packageJsonPath = path.join(filePath, 'package.json')
            if (!fs.existsSync(packageJsonPath)) {
              logger.warn(`no package.json found in ${filePath}`)
              continue
            }

            const json = fs.readJSONSync(packageJsonPath)
            self.resolvePackageJSON(json)
          }
          resolve()
        })
      })
      tasks.push(task)
    }

    await Promise.all(tasks)
    self.checkPackageVersion()
  }

  protected resolvePackageJSON(json: PackageJSON): void {
    const self = this
    self.packageMap.set(json.name, json.version)
    const keys: (keyof PackageJSON)[] = [
      'bundledDependencies',
      'dependencies',
      'devDependencies',
      'optionalDependencies',
      'peerDependencies',
    ]

    const dependencies: PackageDependencies = {}
    for (const key of keys) {
      const currentDependencies = json[key] as PackageDependencies
      if (currentDependencies == null) continue
      for (const [dependency, version] of Object.entries(currentDependencies)) {
        if (dependencies[dependency] == null) {
          dependencies[dependency] = version
          continue
        }
        if (dependencies[dependency] !== version) {
          logger.error(
            `[${json.name}] ${dependency} has difference versions:`,
            `(${dependencies[dependency]}), (${version})`,
          )
          process.exit(-1)
        }
      }
    }

    let hasError = false
    for (const [dependency, version] of Object.entries(dependencies)) {
      const dependencyItem: DependencyItem | undefined = self.dependencyMap.get(
        dependency,
      )

      // new dependency
      if (dependencyItem == null) {
        self.dependencyMap.set(dependency, {
          version,
          dependents: [json.name],
        })
        continue
      }

      // Consistent version number of the same dependency in other packages.
      if (dependencyItem.version === version) {
        dependencyItem.dependents.push(json.name)
        continue
      }

      // eslint-disable-next-line max-len
      // Inconsistent version number of the same dependency in other packages found.
      logger.error(
        `[${json.name}] ${dependency} has a different version(${version}) in other packages:`,
        dependencyItem.dependents
          .map(d => `${d}(${dependencyItem.version})`)
          .join(','),
      )
      hasError = true
    }

    // check for errors
    checkFatalError(hasError)
  }

  protected checkPackageVersion(): void {
    const self = this
    let hasError = false
    for (const [packageName, version] of self.packageMap.entries()) {
      const dependencyItem = self.dependencyMap.get(packageName)
      if (dependencyItem == null) continue

      // There is no '^' prefix of the field 'version' in package.version
      const normalizedDependencyVersion = dependencyItem.version.replace(
        /^[^]/,
        '',
      )
      if (normalizedDependencyVersion === version) continue

      logger.error(
        `[${packageName}] has a different version(${version}) in other packages:`,
        dependencyItem.dependents
          .map(d => `${d}(${dependencyItem.version})`)
          .join(','),
      )
      hasError = true
    }
    checkFatalError(hasError)
  }
}
