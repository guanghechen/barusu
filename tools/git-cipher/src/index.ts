import './env/constant'
export * from './command/decrypt'
export * from './command/encrypt'
export * from './command/init'
export * from './core/command'
export * from './core/decrypt/command'
export * from './core/decrypt/context'
export * from './core/decrypt/processor'
export * from './core/encrypt/command'
export * from './core/encrypt/context'
export * from './core/encrypt/processor'
export * from './core/init/command'
export * from './core/init/context'
export * from './core/init/processor'
export * from './core/option'
export * from './env/constant'
export * from './env/constant'
export * from './env/logger'
export * from './env/util'
export * from './util/buffer'
export * from './util/catalog'
export * from './util/cipher'
export * from './util/events'
export * from './util/io'
export * from './util/secret'
