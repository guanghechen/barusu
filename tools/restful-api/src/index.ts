import './env/constant'

export * from './command/generate'
export * from './command/init'
export * from './command/serve'
export * from './core/command'
export * from './core/generate/command'
export * from './core/generate/context'
export * from './core/generate/processor'
export * from './core/generate/util'
export * from './core/init/command'
export * from './core/init/context'
export * from './core/init/processor'
export * from './core/init/util'
export * from './core/option'
export * from './core/serve/command'
export * from './core/serve/context'
export * from './core/serve/middleware/access-log'
export * from './core/serve/middleware/serve-data-file'
export * from './core/serve/processor'
export * from './env/constant'
export * from './env/logger'
export * from './env/util'
export * from './types/api'
export * from './types/api-item'
export * from './types/api-item-group'
export * from './types/raw/api'
export * from './types/raw/api-item'
export * from './types/raw/api-item-group'
export * from './util/api-parser'
