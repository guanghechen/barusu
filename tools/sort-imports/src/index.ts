import a, * as waw from './env/constant'
import {
  ModuleRankItem,
  StaticImportOrExportStatItem,
  compareModulePath,
  createCommentRegex,
  createStaticImportOrExportRegexList,
  defaultModuleRankItems,
  formatImportOrExportStatItem,
  execWithMultipleRegex,
} from './util'
export * as x from'./env/constant'
export * from './env/logger'
export * from './util'


export class StaticImportStatement {
  public readonly quote: string
  public readonly indent: string
  public readonly semicolon: boolean
  public readonly maxColumn: number
  public readonly itemRank: Record<StaticImportOrExportStatItem['type'], number>
  public readonly moduleRanks: ModuleRankItem[]
  public readonly staticImportOrExportRegexList: RegExp[]
  public readonly topCommentRegex: RegExp

  public constructor(
    quote = '\'',
    indent = '  ',
    semicolon = false,
    maxColumn = 100,
    moduleRanks: ModuleRankItem[] = defaultModuleRankItems.concat(),
    staticImportOrExportRegexList: RegExp[] = createStaticImportOrExportRegexList('g'),
    topCommentRegex: RegExp = createCommentRegex(),
  ) {
    for (let i = 0; i < staticImportOrExportRegexList.length; ++i) {
      const regex = staticImportOrExportRegexList[i]
      if (regex.flags.indexOf('g') < 0) {
        // eslint-disable-next-line no-param-reassign
        staticImportOrExportRegexList[i] = new RegExp(regex, regex.flags + 'g')
      }
    }

    this.quote = quote
    this.indent = indent
    this.semicolon = semicolon
    this.maxColumn = Number.isNaN(maxColumn) ? 100 : maxColumn,
    this.staticImportOrExportRegexList = staticImportOrExportRegexList
    this.topCommentRegex = new RegExp('^(' + topCommentRegex.source + '|\\s*)*')
    this.itemRank = {
      'import': 1,
      'export': 2,
    }
    this.moduleRanks = moduleRanks
  }

  public process(content: string): string {
    const self = this
    const items: StaticImportOrExportStatItem[] = []
    const m = self.topCommentRegex.exec(content)
    const firstNonCommentIndex = m == null ? 0 : m[0].length

    let startIndex = firstNonCommentIndex
    while (true) {
      const execResult = execWithMultipleRegex([], content, startIndex)
      if (execResult  == null) break
      const m: RegExpExecArray = execResult.result
      const regex = execResult.regex

      if (!/^[;\s]*$/.test(content.substring(startIndex, m.index))) break
      const { defaultExport, exportN: exportNStr, moduleName, type, remainOfLine } = m.groups!
      const exportN: string[] = exportNStr == null ? [] : exportNStr.split(/\s*,\s*/g)
      const item = {
        defaultExport,
        exportN: exportN.filter(x => /\S/.test(x)).sort(),
        moduleName: moduleName.replace(/([\/\\])\.[\/\\]/g, '$1').replace(/([\/\\])+/g, '$1'),
        type: type as any,
        remainOfLine: remainOfLine as string,
      } as StaticImportOrExportStatItem
      item.fullStatement = self.format(item)
      items.push(item)
      startIndex = regex.lastIndex
    }

    if (startIndex <= firstNonCommentIndex) return content.trimLeft()
    const newStatements: string = items
      .sort(self.compare.bind(self))
      .map(item => item.fullStatement).join('\n')
    return (
      content.slice(0, firstNonCommentIndex)
      + (newStatements + '\n\n\n' + content.slice(startIndex).trimLeft()).trimRight()
      + '\n'
    )
  }

  protected format(item: Omit<StaticImportOrExportStatItem, 'fullStatement'>): string {
    const { quote, indent, semicolon, maxColumn } = this
    return formatImportOrExportStatItem(item, quote, indent, semicolon, maxColumn)
  }

  protected compare(x: StaticImportOrExportStatItem, y: StaticImportOrExportStatItem): number {
    if (x.type !== y.type) {
      return this.itemRank[x.type] - this.itemRank[y.type]
    }
    if (x.moduleName !== y.moduleName) {
      return compareModulePath(x.moduleName, y.moduleName, this.moduleRanks)
    }
    if (x.fullStatement === y.fullStatement) return 0
    return x.fullStatement < y.fullStatement ? -1 : 1
  }
}
