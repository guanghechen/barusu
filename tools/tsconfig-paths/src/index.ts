import './env/constant'
export * from './command/main'
export * from './core/command'
export * from './core/main/command'
export * from './core/main/context'
export * from './core/main/processor'
export * from './core/main/processor'
export * from './core/option'
export * from './env/constant'
export * from './env/logger'
export * from './util/module-path'
