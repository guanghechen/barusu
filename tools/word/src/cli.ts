import { createProgram, mountSubCommandStat } from '.'

const program = createProgram()

// mount sub-command: stat
mountSubCommandStat(program)

program.parse(process.argv)
