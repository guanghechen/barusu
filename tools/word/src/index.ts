import './env/constant'

export * from './command/stat'
export * from './core/command'
export * from './core/option'
export * from './core/stat/command'
export * from './core/stat/context'
export * from './core/stat/processor'
export * from './env/constant'
export * from './env/logger'
export * from './util/character-stats'
