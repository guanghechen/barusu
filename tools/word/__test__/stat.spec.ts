import { createLoggerMocker } from '@barusu/util-jest'
import fs from 'fs-extra'
import path from 'path'
import { COMMAND_NAME, createProgram, execSubCommandStat, logger } from '../src'

describe('stat', function () {
  const caseRootDirectory = path.resolve(__dirname, 'cases', 'files')
  const workspaceRootDir: string = path.resolve(__dirname, '..')

  const kases = fs.readdirSync(caseRootDirectory)
  for (const kase of kases) {
    const { name: title } = path.parse(kase)
    const projectDir = caseRootDirectory
    const filepath = kase

    // eslint-disable-next-line jest/valid-title
    test(title, async function () {
      const loggerMock = createLoggerMocker({ logger, workspaceRootDir })
      loggerMock.mock()

      const program = createProgram()
      const args = [
        '',
        COMMAND_NAME,
        'stat',
        projectDir,
        '-f',
        filepath,
        '--log-level=debug',
      ]

      await execSubCommandStat(program, args)

      expect(loggerMock.data()).toMatchSnapshot('log info')
      loggerMock.restore()
    })
  }
})
