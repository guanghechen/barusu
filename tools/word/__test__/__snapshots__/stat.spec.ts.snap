// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stat chou_nu_er: log info 1`] = `
Array [
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mcwd: <workspaceRootDir>[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mworkspace: <workspaceRootDir>/__test__/cases/files[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mconfigPath: [][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mparasticConfigPath: <workspaceRootDir>/package.json[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mparasticConfigEntry: @barusu/tool-word[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mencoding: utf-8[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mfilePath: [[39m
[38;2;188;162;31m  \\"<workspaceRootDir>/__test__/cases/files/chou_nu_er.md\\"[39m
[38;2;188;162;31m][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mfilePattern: [][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mshowDetailsPretty: false[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mshowDetails: 0[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mshowSummaryOnly: false[39m
[38;2;188;162;31m[39m",
  ],
  Array [],
  Array [
    "__test__/cases/files/chou_nu_er.md",
  ],
  Array [
    "======================================================
                   total: 76
             blank total:  7
       punctuation total: 12
            unique total: 49
      unique blank total:  1
unique punctuation total:  6
",
  ],
]
`;

exports[`stat ru_meng_ling: log info 1`] = `
Array [
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mcwd: <workspaceRootDir>[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mworkspace: <workspaceRootDir>/__test__/cases/files[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mconfigPath: [][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mparasticConfigPath: <workspaceRootDir>/package.json[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mparasticConfigEntry: @barusu/tool-word[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mencoding: utf-8[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mfilePath: [[39m
[38;2;188;162;31m  \\"<workspaceRootDir>/__test__/cases/files/ru_meng_ling.md\\"[39m
[38;2;188;162;31m][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mfilePattern: [][39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mshowDetailsPretty: false[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mshowDetails: 0[39m
[38;2;188;162;31m[39m",
  ],
  Array [
    "[90m$1<Date>[39m [[38;2;188;162;31mdebug[39m [90mbarusu-word stat[39m]: [38;2;188;162;31mshowSummaryOnly: false[39m
[38;2;188;162;31m[39m",
  ],
  Array [],
  Array [
    "__test__/cases/files/ru_meng_ling.md",
  ],
  Array [
    "======================================================
                   total: 109
             blank total:   7
       punctuation total:  18
            unique total:  70
      unique blank total:   2
unique punctuation total:   5
",
  ],
]
`;
